File: src/js_lsr/lsrattack.js
const SECRET0 = 0;
const SECRET1 = 1;
var x = 0;
var y = 0;
var z = 0;
var obj = {};
obj.x = 0;
obj.y = 0;
obj.z = 0;
var sab = new ArrayBuffer(16);
var arr = new Int32Array(sab);

function lsrattackA0() {
  x = 1;
  if(false) x = SECRET0;
  if(y > 0) { z = 0x12345; } else { z = 0x23456; }
  return x + y + z
}

function lsrattackA1() {
  x = 1;
  if(false) x = SECRET1;
  if(y > 0) { z = 0x12345; } else { z = 0x23456; }
  return x + y + z
}

function lsrattackB0() {
  obj.x = 1;
  if(false) obj.x = SECRET0;
  if(obj.y > 0) { obj.z = 0x12345; } else { obj.z = 0x23456; }
  return obj.x + obj.y + obj.z
}

function lsrattackB1() {
  obj.x = 1;
  if(false) obj.x = SECRET1;
  if(obj.y > 0) { obj.z = 0x12345; } else { obj.z = 0x23456; }
  return obj.x + obj.y + obj.z
}

function lsrattackC0() {
  arr[1] = 1;
  if(false) arr[1] = SECRET0;
  if(arr[2] > 0) { arr[3] = 0x12345; } else { arr[3] = 0x23456; }
  return arr[1] + arr[2] + arr[3]
}

function lsrattackC1() {
  arr[1] = 1;
  if(false) arr[1] = SECRET1;
  if(arr[2] > 0) { arr[3] = 0x12345; } else { arr[3] = 0x23456; }
  return arr[1] + arr[2] + arr[3]
}

function callVeryRecognizableFunctions() {
  var a = 0;
  for(i = 0; i < 1000000; i++) {
    a += lsrattackA0();
    y = x;
    obj.y = obj.x;
    arr[2] = arr[1];

    a += lsrattackA1();
    y = x;
    obj.y = obj.x;
    arr[2] = arr[1];

    a += lsrattackB0();
    y = x;
    obj.y = obj.x;
    arr[2] = arr[1];

    a += lsrattackB1();
    y = x;
    obj.y = obj.x;
    arr[2] = arr[1];

    a += lsrattackC0();
    y = x;
    obj.y = obj.x;
    arr[2] = arr[1];

    a += lsrattackC1();
    y = x;
    obj.y = obj.x;
    arr[2] = arr[1];
  }
  print(a)
}

callVeryRecognizableFunctions()

File: src/java_lsr/LSRAttack.java
public class LSRAttack {
  public static void main(String[] args) {
    new LSRAttack().attack();
  }

  public void attack() {
    int total = 0;
    for(int i = 0; i < 1000000; i++) {
      total += lsrattackA0(true);
      total += lsrattackA1(true);
      total += lsrattackA0_const(true);
      total += lsrattackA1_const(true);
      total += lsrattackB0(true);
      total += lsrattackB1(true);
      total += lsrattackC0(true);
      total += lsrattackC1(true);
    }
    System.out.println("Hello world " + total);
  }

  private static int SECRET0 = 0;
  private static int SECRET1 = 1;

  public static final boolean alwaysFalse_const = false;

  private static int x_static, y_static, z_static;
  private int x_inst, y_inst, z_inst;

  public static int lsrattackA0(boolean withPrivileges) {
    x_static = 1;
    if(withPrivileges) x_static = SECRET0;
    if(y_static > 0) { z_static = 0x12345; } else { z_static = 0x23456; }
    return x_static + y_static + z_static;
  }
  public static int lsrattackA1(boolean withPrivileges) {
    x_static = 1;
    if(withPrivileges) x_static = SECRET1;
    if(y_static > 0) { z_static = 0x12345; } else { z_static = 0x23456; }
    return x_static + y_static + z_static;
  }

  public static int lsrattackA0_const(boolean withPrivileges) {
    x_static = 1;
    if(alwaysFalse_const) x_static = SECRET0;
    if(y_static > 0) { z_static = 0x12345; } else { z_static = 0x23456; }
    return x_static + y_static + z_static;
  }
  public static int lsrattackA1_const(boolean withPrivileges) {
    x_static = 1;
    if(alwaysFalse_const) x_static = SECRET1;
    if(y_static > 0) { z_static = 0x12345; } else { z_static = 0x23456; }
    return x_static + y_static + z_static;
  }

  public int lsrattackB0(boolean withPrivileges) {
    x_inst = 1;
    if(withPrivileges) x_inst = SECRET0;
    if(y_inst > 0) { z_inst = 0x12345; } else { z_inst = 0x23456; }
    return x_inst + y_inst + z_inst;
  }
  public int lsrattackB1(boolean withPrivileges) {
    x_inst = 1;
    if(withPrivileges) x_inst = SECRET1;
    if(y_inst > 0) { z_inst = 0x12345; } else { z_inst = 0x23456; }
    return x_inst + y_inst + z_inst;
  }

  public int lsrattackC0(boolean withPrivileges) {
    if(withPrivileges) x_inst = SECRET0;
    else x_inst = 1;
    if(y_inst > 0) { z_inst = 0x12345; } else { z_inst = 0x23456; }
    return x_inst + y_inst + z_inst;
  }
  // HotSpot (trained on withPrivileges=TRUE):
  // if(!withPrivileges) bail
  // r = y_inst;
  // x_inst = SECRET0;
  // if(r > 0) bail
  // z_inst = 0x23456;
  // ...
  public int lsrattackC1(boolean withPrivileges) {
    if(withPrivileges) x_inst = SECRET1;
    else x_inst = 1;
    if(y_inst > 0) { z_inst = 0x12345; } else { z_inst = 0x23456; }
    return x_inst + y_inst + z_inst;
  }
  // HotSpot (trained on withPrivileges=TRUE):
  // if(withPrivileges) x_inst = SECRET1;
  // else x_inst = 1;
  // if(y_inst > 0) z_inst = 0x12345;
  // else z_inst = 0x23456;
  // ...
}
